yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -825,6 +813,9 @@ private[spark] class Client(
    * throw an appropriate SparkException.
    */
   def run(): Unit = {
+    // Cleanup staging director as some appStagingDir can not be deleted when job is failed or
+    // killed, please see SPARK-7705 for details.
+    cleanupStagingDir()
```

Comment:
Wait, why is this being called before the YARN application even runs? Does this even work on a real cluster?